Source/Storages/Cloud/CloudKeyManager.swift

```diff
@@ -134,3 +219,69 @@ extension CloudKeyManager {
         }
     }
 }
+
+// MARK: Named Keys
+
+private extension CloudKeyManager {
+    func store(key: VirgilPrivateKey, keyName: String, usingPassword password: String) throws {
+        let exportedIdentityKey = try self.crypto.exportPrivateKey(key)
+
+        let brainKeyPair = try self.brainKey.generateKeyPair(password: password).startSync().get()
+
+        let keyknoxValue = try self.pullKeyValue(named: keyName, brainKeyPair: brainKeyPair)
+
+        guard keyknoxValue.value.isEmpty, keyknoxValue.meta.isEmpty else {
+            throw CloudKeyStorageError.entryAlreadyExists
+        }
+
+        let now = Date()
+        let cloudEntry = CloudEntry(name: self.identity,
+                                    data: exportedIdentityKey,
+                                    creationDate: now,
+                                    modificationDate: now,
+                                    meta: nil)
+
+        let data = try JSONEncoder().encode(cloudEntry)
+
+        try self.pushKeyValue(named: keyName, data: data, hash: keyknoxValue.keyknoxHash, brainKeyPair: brainKeyPair)
+    }
+
+    func retrieve(usingPassword password: String, keyName: String) throws -> CloudEntry {
+        let brainKeyPair = try self.brainKey.generateKeyPair(password: password).startSync().get()
+
+        do {
+            let keyknoxValue = try self.pullKeyValue(named: keyName, brainKeyPair: brainKeyPair)
+
+            return try JSONDecoder().decode(CloudEntry.self, from: keyknoxValue.value)
+        }
+        catch KeyknoxCryptoError.decryptionFailed {
+            throw EThreeError.wrongPassword
+        }
+    }
+
+    private func delete(keyName: String) throws {
+        let params = KeyknoxResetParams(root: self.namedKeysRoot, path: namedKeysPath, key: keyName)
+        _ = try self.keyknoxManager.resetValue(params: params).startSync().get()
+    }
+
+    func changePassword(from oldPassword: String,
+                        to newPassword: String,
+                        keyName: String) throws {
+        let brainKeyPair = try self.brainKey.generateKeyPair(password: oldPassword).startSync().get()
+
+        let keyknoxValue = try self.pullKeyValue(named: keyName, brainKeyPair: brainKeyPair)
+
+        guard !keyknoxValue.value.isEmpty, !keyknoxValue.meta.isEmpty else {
+            return
+        }
+
+        sleep(2)
```

Comment:
Why we need a 2 seconds delay?

Reply:
Due to the throttling of the requests for generating brain key on virgil pythia server. Were there any changes with this problem?

Reply:
Is there 2 calls to the `brain key` service followed one by one?

Reply:
Yes. Lines 270 and 280 - uses `BrainKey.generateKeyPair(password:)` - which calls to brain key service